lib/phoenix_live_view/html_engine.ex

```diff
@@ -435,9 +435,21 @@ defmodule Phoenix.LiveView.HTMLEngine do
     attrs = remove_phx_no_break(attrs)
     validate_phx_attrs!(attrs, tag_meta, state)
 
-    state
-    |> set_root_on_tag()
-    |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
+    case pop_special_attr!(attrs, "for", state) do
+      {{":for", expr, _meta}, attrs} ->
+        ast =
+          state
+          |> set_root_on_not_tag()
+          |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
+          |> handle_for_expr(expr)
+
+        update_subengine(state, :handle_expr, ["=", ast])
```

Comment:
You still need to call handle_begin in this case:

```suggestion
        state
        |> update_subengine(:handle_begin, [])
        |> set_root_on_not_tag()
        |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
        |> handle_for_expr(expr, state)
```

Then you can pass the old state to `handle_for_expr(...)`. `handle_for_expr()` can go ahead and call `        update_subengine(state, :handle_expr, ["=", ast])` on the old state too. :)